zebra-chain/src/amount.rs

```diff
@@ -191,6 +191,60 @@ where
     }
 }
 
+impl<C> std::ops::Mul<u64> for Amount<C>
+where
+    C: Constraint,
+{
+    type Output = Result<Amount<C>>;
+
+    fn mul(self, rhs: u64) -> Self::Output {
+        let value = (self.0 as u64)
+            .checked_mul(rhs)
+            .ok_or(Error::AmountMulError)?;
+        value.try_into()
+    }
+}
+
+impl<C> std::ops::Mul<Amount<C>> for u64
+where
+    C: Constraint,
+{
+    type Output = Result<Amount<C>>;
+
+    fn mul(self, rhs: Amount<C>) -> Self::Output {
+        rhs.mul(self)
+    }
+}
+
+#[derive(thiserror::Error, Debug, displaydoc::Display, Clone, PartialEq)]
+#[allow(missing_docs)]
+pub enum AmountDivError {
+    /// division by zero is an invalid operation, input {value}
+    DivideByZero { value: i64 },
+}
+
+impl From<AmountDivError> for Error {
+    fn from(err: AmountDivError) -> Error {
+        Error::AmountDivError(err)
+    }
+}
+
+impl<C> std::ops::Div<u64> for Amount<C>
+where
+    C: Constraint,
+{
+    type Output = Result<Amount<C>>;
+    fn div(self, rhs: u64) -> Self::Output {
+        let quotient = self
+            .0
+            .checked_div(rhs as i64)
+            .ok_or(AmountDivError::DivideByZero { value: rhs as i64 })?;
+        // since this is a division by a positive integer,
+        // the quotient must be within the constrained range
+        Ok(quotient.try_into().unwrap())
+    }
+}
```

Comment:
We can't convert rhs to i64, because it might overflow.

Instead, convert self to u64, which must succeed, because Amount is NonNegative.

```suggestion
impl std::ops::Div<u64> for Amount<NonNegative>
{
    type Output = Result<Amount<NonNegative>>;
    
    fn div(self, rhs: u64) -> Self::Output {
        let quotient = (self.0 as u64)
            .checked_div(rhs)
            .ok_or(DivideByZero { amount: self.0 })?;
        // since this is a division by a positive integer,
        // the quotient must be within the constrained range
        Ok(quotient.try_into().unwrap())
    }
}
```